### PR TITLE
review elyses enchantments

### PR DESCRIPTION
**Pull request by jiff-korakoch**

### CODE REVIEW

File: elyses-enchantments/enchantments.js [hunk 1]
```diff
@@ -22,8 +22,9 @@ export function getItem(cards, position) {
  * @returns {number[]} the cards with the change applied
  */
 export function setItem(cards, position, replacementCard) {
-  cards[position] = replacementCard;
-  return cards;
+  let newCards = Array.from(cards);
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: We can use the _spread_ [syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) when cloning arrays.

```js
  let newCards = [ ...cards ];
```

Comment (gomain) [inline elyses-enchantments/enchantments.js]: Also can use `const` here.

File: elyses-enchantments/enchantments.js [hunk 2]
```diff
@@ -35,8 +36,9 @@ export function setItem(cards, position, replacementCard) {
  * @returns {number[]} the cards with the newCard applied
  */
 export function insertItemAtTop(cards, newCard) {
-  cards.push(newCard);
-  return cards;
+  const newCards = Array.from(cards);
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: ditto

File: elyses-enchantments/enchantments.js [hunk 3]
```diff
@@ -48,8 +50,9 @@ export function insertItemAtTop(cards, newCard) {
  * @returns {number[]} the cards without the removed card
  */
 export function removeItem(cards, position) {
-  cards.splice(position, 1);
-  return cards;
+  const newCards = Array.from(cards);
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: ditto

File: elyses-enchantments/enchantments.js [hunk 4]
```diff
@@ -60,8 +63,9 @@ export function removeItem(cards, position) {
  * @returns {number[]} the cards without the removed card
  */
 export function removeItemFromTop(cards) {
-  cards.pop();
-  return cards;
+  const newCards = Array.from(cards);
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: ditto

File: elyses-enchantments/enchantments.js [hunk 5]
```diff
@@ -73,8 +77,9 @@ export function removeItemFromTop(cards) {
  * @returns {number[]} the cards including the new card
  */
 export function insertItemAtBottom(cards, newCard) {
-  cards.unshift(newCard);
-  return cards;
+  const newCards = Array.from(cards);
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: ditto

File: elyses-enchantments/enchantments.js [hunk 6]
```diff
@@ -35,8 +36,9 @@ export function setItem(cards, position, replacementCard) {
  * @returns {number[]} the cards with the newCard applied
  */
 export function insertItemAtTop(cards, newCard) {
-  cards.push(newCard);
-  return cards;
+  const newCards = Array.from(cards);
+  newCards.push(newCard);
+  return newCards;
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: When used with spread syntax,
```js
  return [ ...cards, newCard ];
```

File: elyses-enchantments/enchantments.js [hunk 7]
```diff
@@ -48,8 +50,9 @@ export function insertItemAtTop(cards, newCard) {
  * @returns {number[]} the cards without the removed card
  */
 export function removeItem(cards, position) {
-  cards.splice(position, 1);
-  return cards;
+  const newCards = Array.from(cards);
+  newCards.splice(position, 1);
+  return newCards;
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: Try to think of a way that does not mutate arrays.

```js
  return [ ...<items before removal>, ...<items after removal> ];
```

See [slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice)

**Review by gomain — CHANGES_REQUESTED** [review]
Pleas look into how we might not need to mutate arrays at all.

I.e. always read from the received array in constructing the returned array.

File: elyses-enchantments/enchantments.js [hunk 8]
```diff
@@ -22,7 +22,7 @@ export function getItem(cards, position) {
  * @returns {number[]} the cards with the change applied
  */
 export function setItem(cards, position, replacementCard) {
-  let newCards = Array.from(cards);
+  const newCards = [...cards];
   newCards[position] = replacementCard;
   return newCards;
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: `newCards` is still mutated.

File: elyses-enchantments/enchantments.js [hunk 9]
```diff
@@ -50,9 +48,7 @@ export function insertItemAtTop(cards, newCard) {
  * @returns {number[]} the cards without the removed card
  */
 export function removeItem(cards, position) {
-  const newCards = Array.from(cards);
-  newCards.splice(position, 1);
-  return newCards;
+  return [...cards.slice(-cards.length,position),...cards.slice(position+1)];
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: Parameters of `Array#slice` are `indexFrom` and `indexTo` (exclusive).

```js
const arr = [ 1, 2, 3, 4, 5 ]; // => <0> 1 <1> 2 <2> 3 <3> 4 <4> 5
arr.slice(2, 4) // =>                             ^           ^    => [ 3, 4 ]

// not want <2>                       ^-----------^     ^...............
const no2 = [...arr.slice(0, 2), ...arr.slice(3)];
//                                   <-5>                    <-1>
```
Why use `-cards.length` ?

File: elyses-enchantments/enchantments.js (hunk 8 above)
Comment (gomain) [reply elyses-enchantments/enchantments.js]: Look at `removeItem`.

File: elyses-enchantments/enchantments.js [hunk 10]
```diff
@@ -63,7 +59,7 @@ export function removeItem(cards, position) {
  * @returns {number[]} the cards without the removed card
  */
 export function removeItemFromTop(cards) {
-  const newCards = Array.from(cards);
+  const newCards = [...cards];
   newCards.pop();
   return newCards;
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: Still mutating `newCards`

File: elyses-enchantments/enchantments.js [hunk 11]
```diff
@@ -77,7 +73,7 @@ export function removeItemFromTop(cards) {
  * @returns {number[]} the cards including the new card
  */
 export function insertItemAtBottom(cards, newCard) {
-  const newCards = Array.from(cards);
+  const newCards = [...cards];
   newCards.unshift(newCard);
   return newCards;
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: ditto

File: elyses-enchantments/enchantments.js [hunk 12]
```diff
@@ -90,7 +86,7 @@ export function insertItemAtBottom(cards, newCard) {
  * @returns {number[]} the cards without the removed card
  */
 export function removeItemAtBottom(cards) {
-  const newCards = Array.from(cards);
+  const newCards = [...cards];
   newCards.shift();
   return newCards;
```
Comment (gomain) [inline elyses-enchantments/enchantments.js]: ditto

**Review by gomain — CHANGES_REQUESTED** [review]
Get it?

File: elyses-enchantments/enchantments.js (hunk 1 above)
Comment (jiff-korakoch) [reply elyses-enchantments/enchantments.js]: @gomain fixed.

**Review by gomain — CHANGES_REQUESTED** [review]
Please fix styles.

Arrays:
```js
const arr = [1, 2, 3, ...spreading];
//           ^                    ^  no space
//              ^     ^              space after item
```
Parameter list:
```js
function name(name1, name2) { ...
//                   ^      space after item

const result = name(param1, param2);
//                          ^       space after item
```
If line gets too long (>80):
```js
const arr = [
  item1,
  item2,
  item3,
  item4, // comma after every item
];

function name(
  param1,
  param2,
  param3,
  param4, // comma after every parameter
) { ...   // body on next line
}

const result = name(
  arg1,
  arg2,
  arg3,
  arg4, // comma after every argument
);

```
Space around operators
```js
const result = 4 + 2 // not 4+2
name(param + 1); // not `param+1`
```

**Review by gomain — APPROVED** [review]
very good

**Pull request merged**